README.md
# ![angular-webcolor][.svg] Angular-webcolor [![Bower version][bower-image]][bower] [![Build Status][travis-image]][travis]

* [DEMO](http://jsrun.it/59naga/angular-webcolor)
* [DEMO2](http://jsrun.it/59naga/yHIb)

## Installation
```bash
$ bower install angular angular-ui-router angular-webcolor
```

```html
<html ng-app="myApp">
  <head>
  <script src="bower_components/angular/angular.min.js"></script>
  <script src="bower_components/angular-ui-router/release/angular-ui-router.min.js"></script>
  <script src="bower_components/angular-webcolor/public/angular-webcolor.min.js"></script>
  <script>
    angular
    .module('myApp',['ui.router','webcolor'])
    .config(function($stateProvider){
      $stateProvider.state('second',{
        url:'second',
        template:'{{message}} <br><a ui-sref="first">first</a>',
        controller:function($scope){
          $scope.message= 'this is second'
        },
      });
      $stateProvider.state('first',{
        url:'*path',
        resolve:{
          delay:function($timeout){
            return $timeout(function(){return 0},1000)
          },
        },
        template:'{{message}} <br><a ui-sref="second">second</a>',
        controller:function($scope,delay){
          $scope.message= 'this is first'
        },
      });
    })
    .run(function($rootScope,$webcolorLoadingBar){
      $rootScope.$on('$stateChangeStart',function(){
        $webcolorLoadingBar.start();
      });
      $rootScope.$on('$stateChangeSuccess',function(){
        $webcolorLoadingBar.complete();
      });
    });
  </script>
  </head>
  <body ui-view>loading...</body>  
</html>
```

# API
## webcolors
is angular.constant Array the **[140 webcolor names](http://www.w3schools.com/html/html_colornames.asp)**.

example:
```js
app.config(function(webcolors){
  console.log(webcolors); // Array[140] "aliceblue","antiquewhite","aqua","...""
})
```
## $webcolorLoadingBar
### .start()
Begin render.
### .complete()
Fast forward render.

### Change $webcolorLoadingBar behavior
```js
angular
.module('myApp',['ui.router','webcolor'])
.config(function($webcolor){
  $webcolor.delay= 50;// default 100
  $webcolor.opacity= 1;// default .5

  $webcolor.zIndex= 50;// default 1000
  $webcolor.lines= 100;// default 1

  $webcolor.begin= 1000;// default 0
  $webcolor.endIncrement= 3;// default 2
  $webcolor.endTranslucent= 0.02;// default 0.0125
})
```
### $webcolor.delay
(msec) Rendering speed of 1 pixel.

### $webcolor.opacity
(0~1) Set the initial opacity.

### $webcolor.zIndex
(0~) Set css z-index property for $webcolorLoadingBar.

### $webcolor.lines
(0~100) Number of $webcolorLoadingBar

### $webcolor.begin
(msec) Set the maximum randam delay for begin render.

### $webcolor.endIncrement
(int) Rendering number of pixel for delay. after $webcolorLoadingBar.complete();
### $webcolor.endTranslucent
(float) Set transclunet per frame. after rendered.

# License
[MIT][License] by @59naga

[License]: http://59naga.mit-license.org/

[.svg]: https://cdn.rawgit.com/59naga/angular-webcolor/master/.svg?

[bower-image]: https://badge.fury.io/bo/angular-webcolor.svg
[bower]: http://badge.fury.io/bo/angular-webcolor
[travis-image]: https://travis-ci.org/59naga/angular-webcolor.svg?branch=master
[travis]: https://travis-ci.org/59naga/angular-webcolor


CHANGELOG.md
v0.0.18 / Apr 26 2015
=========================
 * [`unknown`][4] :bug: Fix [#2][4A]

[4]: https://github.com/59naga/angular-webcolor/commits/master
[4A]: https://github.com/59naga/angular-webcolor/issues/2

v0.0.17 / Apr 12 2015
=========================
 * [`811f8d4`][2] :bug: Fix minus opacity for safari
 * [`354cb64`][3] :art: Add $webcolor constant

[2]: https://github.com/59naga/angular-webcolor/commit/811f8d486593824134c9919c7894f7b18af394fd
[3]: https://github.com/59naga/angular-webcolor/commit/354cb64254d0db095c225f9f4d89ae9f02ddc1d0

v0.0.13 / Mar 29 2015
=========================
 * [`unknown`][1] :lipstick: Refactoring and Test

[1]: https://github.com/59naga/angular-webcolor/commit/8d120aa683e5130d02038f1e1c6cc1b1f26a660a

[0]: https://github.com/59naga/angular-webcolor/commits/master


public/angular-webcolor.js
(function() {
  var Progress, module;

  module = angular.module('webcolor', []);

  module.constant('webcolors', ["aliceblue", "antiquewhite", "aqua", "aquamarine", "azure", "beige", "bisque", "black", "blanchedalmond", "blue", "blueviolet", "brown", "burlywood", "cadetblue", "chartreuse", "chocolate", "coral", "cornflowerblue", "cornsilk", "crimson", "cyan", "darkblue", "darkcyan", "darkgoldenrod", "darkgray", "darkgreen", "darkkhaki", "darkmagenta", "darkolivegreen", "darkorange", "darkorchid", "darkred", "darksalmon", "darkseagreen", "darkslateblue", "darkslategray", "darkturquoise", "darkviolet", "deeppink", "deepskyblue", "dimgray", "dodgerblue", "firebrick", "floralwhite", "forestgreen", "fuchsia", "gainsboro", "ghostwhite", "gold", "goldenrod", "gray", "green", "greenyellow", "honeydew", "hotpink", "indianred", "indigo", "ivory", "khaki", "lavender", "lavenderblush", "lawngreen", "lemonchiffon", "lightblue", "lightcoral", "lightcyan", "lightgoldenrodyellow", "lightgreen", "lightgrey", "lightpink", "lightsalmon", "lightseagreen", "lightskyblue", "lightslategray", "lightsteelblue", "lightyellow", "lime", "limegreen", "linen", "magenta", "maroon", "mediumaquamarine", "mediumblue", "mediumorchid", "mediumpurple", "mediumseagreen", "mediumslateblue", "mediumspringgreen", "mediumturquoise", "mediumvioletred", "midnightblue", "mintcream", "mistyrose", "moccasin", "navajowhite", "navy", "oldlace", "olive", "olivedrab", "orange", "orangered", "orchid", "palegoldenrod", "palegreen", "paleturquoise", "palevioletred", "papayawhip", "peachpuff", "peru", "pink", "plum", "powderblue", "purple", "red", "rosybrown", "royalblue", "saddlebrown", "salmon", "sandybrown", "seagreen", "seashell", "sienna", "silver", "skyblue", "slateblue", "slategray", "snow", "springgreen", "steelblue", "tan", "teal", "thistle", "tomato", "turquoise", "violet", "wheat", "white", "whitesmoke", "yellow", "yellowgreen"]);

  module.constant('$webcolor', {
    delay: 100,
    opacity: .5,
    zIndex: 1000,
    lines: 1,
    begin: 0,
    endIncrement: 2,
    endTranslucent: 0.0125
  });

  module.provider('$webcolorLoadingBar', ["$webcolor", function($webcolor) {
    var progresses;
    progresses = [];
    return {
      $get: ["webcolors", function(webcolors) {
        return {
          start: function() {
            var i, j, k, len, progress, ref, results;
            if (progresses.length) {
              for (j = 0, len = progresses.length; j < len; j++) {
                progress = progresses[j];
                progress.stop();
              }
            }
            progresses = [];
            results = [];
            for (i = k = 0, ref = $webcolor.lines; 0 <= ref ? k < ref : k > ref; i = 0 <= ref ? ++k : --k) {
              progress = new Progress(webcolors, {
                i: i,
                delay: $webcolor.delay,
                zIndex: $webcolor.zIndex,
                opacity: $webcolor.opacity,
                begin: Math.random() * $webcolor.begin,
                endIncrement: $webcolor.endIncrement,
                endTranslucent: $webcolor.endTranslucent
              });
              results.push(progresses.push(progress));
            }
            return results;
          },
          complete: function() {
            var j, len, progress, results;
            results = [];
            for (j = 0, len = progresses.length; j < len; j++) {
              progress = progresses[j];
              results.push(progress.hurry());
            }
            return results;
          }
        };
      }]
    };
  }]);

  Progress = (function() {
    function Progress(colors, options) {
      this.colors = colors;
      this.options = options;
      this.i = 0;
      this.delay = this.options.delay;
      this.opacity = this.options.opacity;
      this.canvas = document.createElement('canvas');
      this.canvas.className = 'webcolor';
      this.canvas.width = window.innerWidth;
      this.canvas.height = Math.ceil(window.innerHeight / 100);
      this.x = 0;
      this.y = this.canvas.height * this.options.i;
      this.context = this.canvas.getContext('2d');
      setTimeout((function(_this) {
        return function() {
          return _this.next();
        };
      })(this), this.options.begin);
    }

    Progress.prototype.hurry = function() {
      return this.delay = 0;
    };

    Progress.prototype.next = function() {
      var fullfilled, i, j, ref;
      if (this.stopped) {
        return;
      }
      fullfilled = innerWidth < (this.i * this.canvas.height);
      if (fullfilled) {
        if (this.opacity <= 0) {
          return this.stop();
        }
        if (this.delay === 0) {
          this.opacity -= this.options.endTranslucent;
        }
      } else {
        this.nextPixel();
        if (this.delay === 0) {
          for (i = j = 1, ref = this.options.endIncrement; 1 <= ref ? j <= ref : j >= ref; i = 1 <= ref ? ++j : --j) {
            this.nextPixel();
          }
        }
      }
      this.updateStyle();
      return setTimeout((function(_this) {
        return function() {
          return requestAnimationFrame(function() {
            return _this.next();
          });
        };
      })(this), this.delay);
    };

    Progress.prototype.stop = function() {
      var ref;
      this.stopped = true;
      if (((ref = this.canvas) != null ? ref.parentNode : void 0) != null) {
        return this.canvas.parentNode.removeChild(this.canvas);
      }
    };

    Progress.prototype.getStyle = function() {
      var opacity;
      opacity = this.opacity;
      if (this.opacity < 0) {
        opacity = 0;
      }
      return ["z-index:" + (parseInt(this.options.zIndex)), "opacity:" + opacity, "position:fixed", "left:" + this.x + "px", "top:" + this.y + "px"].join(';');
    };

    Progress.prototype.updateStyle = function(parentNode) {
      if (parentNode == null) {
        parentNode = 'html';
      }
      parentNode = document.querySelector(parentNode);
      if (parentNode !== this.canvas.parentNode) {
        parentNode.appendChild(this.canvas);
      }
      if (this.canvas.getAttribute('style') !== this.getStyle()) {
        return this.canvas.setAttribute('style', this.getStyle());
      }
    };

    Progress.prototype.nextPixel = function() {
      var contextX;
      contextX = this.i++ * this.canvas.height;
      this.context.fillStyle = this.colors[~~(this.colors.length * Math.random())];
      return this.context.fillRect(contextX, 0, this.canvas.height, this.canvas.height);
    };

    return Progress;

  })();

}).call(this);
